### PR TITLE
First PowerVM backend implementation

### PR DESCRIPTION
**Pull request by k0da**
This is first implementation of a PowerVM backend.

Signed-off-by: Dinar Valeev <dvaleev@suse.com>

### CODE REVIEW

**Comment by aaannz** [comment]
Tidy is failing the travis.

**Comment by k0da** [comment]
FIxed

File: backend/pvm.pm [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    powercontrol('off');
+    runcmd("rmvterm", "--id", $vars->{LPARID});
+    for my $i (1 .. $vars->{NUMDISKS}) {
+        my $disk = $vars->{LPAR} . "_" . $i;
+        $cmd = "$pvmctl scsi delete --type lv --lpar name=$vars->{LPAR} --stor-id=$disk";
+        runcmd($cmd);
+        $cmd = "sudo viosvrcmd --id 1 -c \"rmbdsp -sp rootvg -bd $disk\"";
+        runcmd($cmd);
+    }
+    $cmd = "$pvmctl scsi delete --type vopt --lpar name=$vars->{LPAR} --stor-id=$vars->{VIOISO}";
```
Comment (coolo) [inline backend/pvm.pm]: what bothers me here is that you duplicate this code - and even inconsistently. Here you use a string and in the create case you use an array. Both share a lot of paramters, so possibly they both should inherit a common array.

File: backend/pvm.pm [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    }
+    if ($action eq 'reset') {
+        $cmd = "restart";
+    }
+    runcmd("/usr/bin/pvmctl", "lpar", $cmd, "-i", "name=" . $vars->{LPAR});
+}
+sub attach_console {
+    my ($self, $args) = @_;
+    my $vars    = \%bmwqemu::vars;
+    my $vncport = qx/sudo \/usr\/sbin\/mkvtermutil --id $vars->{LPARID} --vnc --local --log serial0 2>\/dev\/null/;
```
Comment (coolo) [inline backend/pvm.pm]: this evil quoting looks horrible. You can use qx{} too - using / as seperator with a string with tons of / is kind of crazy

**Comment by k0da** [comment]
Comments are addressed.

File: doc/backend_vars.asciidoc [hunk 3]
```diff
@@ -84,3 +84,17 @@ VIRSH_HOSTNAME;string;SSH Host with virtsh;
...
+[options="header",cols="^m,^m,^m,v",separator=";"]
+|====================
+Variable;Values allowed;Default value;Explanation
+MEM;integer;2048;amount of RAM;
+LPAR;string;osauto;LPAR name to be created;
+NUMDISKS;integer;1;Number of disks;
+HDDSIZEGB;integer;15;Disk size in GB;
+NICVLAN;integer;1;VLAN to attach to;
+VSWITCH;string;VSWITCH0;A virtula switch to connect to;
+CPUS;integer;1;Number of CPUS for LPAR;
```
Comment (okurz) [inline doc/backend_vars.asciidoc]: hm, the other backends use a prefix for their variables. Please check again. Do you really also need "MEM" and such?
@aaannz how does this relate to https://github.com/os-autoinst/os-autoinst/issues/339 ?

Comment (aaannz) [inline doc/backend_vars.asciidoc]: Well, qemu backend doesn't use any prefix. Also its variables are not universal across backends. i.e. IPMI certainly doesn't have MEM option, so I am for listing it (and other similar) here.

**Comment by k0da** [comment]
mem is fixed. 

Are vars needs to be prefixed? They are listed in documentation.

File: backend/pvm.pm [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    $vars->{MEM}  ||= "2048";
+    #disk settings
+    $vars->{NUMDISKS}  ||= 1;
+    $vars->{HDDSIZEGB} ||= 15;
+    #network settings
+    $vars->{NIC}     ||= "sea";
+    $vars->{NICVLAN} ||= 1;
+    $vars->{VSWITCH} ||= "ETHERNET0";
+    # generate the ISO name for VIO
+    my $os      = $vars->{DISTRI};
```
Comment (coolo) [inline backend/pvm.pm]: I was about to merge this, but I notice this and I have my doubts this belongs here. Such distribution specific things don't really belong in os-autoinst, left alone in a backend. IMHO this belongs in openQA - set in the medium

Comment (k0da) [inline backend/pvm.pm]: Yes. But this limitation is really stupid and I'm afraid I can't make it fixed. (Those parts are closed source). And I want to keep similarity across architectures, The iso name just changed temporarily in background.

Comment (coolo) [inline backend/pvm.pm]: you can still set the short ISO in openQA - we support variable expansion in the scheduler

Comment (k0da) [inline backend/pvm.pm]: Right, but isotovideo will stop working then.

Comment (coolo) [inline backend/pvm.pm]: isotovideo will require a VIOISO variable in the test - that's not "stop working"

**Comment by k0da** [comment]
Ok, got rid of it. By generating a md5sum of a filename.

File: backend/pvm.pm [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+    my $id = substr(qx/pvmctl lpar list -d LogicalPartition.id --where LogicalPartition.name=$vars->{LPAR}/, 3);
+    chomp($id);
+    $vars->{LPARID} = $id;
+    bmwqemu::save_vars();
+
+    #unfortunately NovaLink can't take files longer than 38 chars. So we need to shorten ISO name here
+    #we copy isos from nfs mount on VIO side to VMLibrary
+    my $source_iso = $vars->{ISO};
+    $source_iso =~ s/var\/lib\/openqa\/share\/factory\///g;
```
Comment (coolo) [inline backend/pvm.pm]: I would feel better if you used 'iso/' . basename($source_iso) - in the light of https://progress.opensuse.org/issues/10470 it sounds bad to hardcode openQA paths in os-autoinst.

Comment (k0da) [inline backend/pvm.pm]: Makes sense, yes. Thanks

**Pull request merged**